elliottlib/imagecfg.py

```diff
@@ -23,32 +24,25 @@ def base_only(self):
         """
         return self.config.base_only
 
-    def get_latest_build_info(self, product_version):
+    def get_latest_build_info(self, product_version, tag_set, latest_builds, brew_session):
         """
         Queries brew to determine the most recently built release of the component
         associated with this image. This method does not rely on the "release"
         label needing to be present in the Dockerfile.
 
         :return: A tuple: (component name, version, release); e.g. ("registry-console-docker", "v3.6.173.0.75", "1")
         """
-
         component_name = self.get_component_name()
-
         tag = "{}-candidate".format(self.branch())
-
-        rc, stdout, stderr = exectools.cmd_gather(["brew", "latest-build", tag, component_name])
-
-        assertion.success(rc, "Unable to search brew builds: %s" % stderr)
-
-        latest = stdout.strip().splitlines()[-1].split(' ')[0]
-
-        if not latest.startswith(component_name):
-            # If no builds found, `brew latest-build` output will appear as:
-            # Build                                     Tag                   Built by
-            # ----------------------------------------  --------------------  ----------------
-            raise IOError("No builds detected for %s using tag: %s" % (self.qualified_name, tag))
-
-        # latest example: "registry-console-docker-v3.6.173.0.75-1""
-        name, version, release = latest.rsplit("-", 2)  # [ "registry-console-docker", "v3.6.173.0.75", "1"]
-
-        return name, version, release, product_version[self.name]
+        if tag not in tag_set:
+            latest_builds[:] = latest_builds + brew_session.getLatestBuilds(tag)
+            tag_set.add(tag)
+
+        for b in latest_builds:
+            nvr = b.get("nvr", "")
+            if nvr.startswith(component_name):
+                name, version, release = nvr.rsplit("-", 2)
+                return name, version, release, product_version[self.name]
```

Comment:
Suppose we have `M` images in `image_metadata` and `N` builds returned by each call of `getLatestBuilds`,  eventually:

1.  The loop will run `O(M*N)` times. 
2.  Suppose the component names have an average length `L`, each call of `nvr. startswith` will cost `O(L)` time. 

Then the overall time complexity of retrieving latest NVRs is `O(M*N*L)`.

Reply:
Yeah, I aware of that. I can parse the API return list of builds and create a dict key base on substr of the nvr. then it will reduced to O(1) (for this looping part) is that what you want?

Reply:
Yeah, your proposed idea is really nice.